program-runtime/src/prioritization_fee.rs

```diff
@@ -0,0 +1,183 @@
+// Prioritization fee rate is 1 lamports per 10K CUs
+const COMPUTE_UNIT_TICK_SIZE: u64 = 10_000;
+
+pub enum PrioritizationFeeType {
+    Rate(u64),
+    Deprecated(u64),
+}
+
+#[derive(Default, Debug, PartialEq)]
+pub struct PrioritizationFeeDetails {
+    fee: u64,
+    priority: u64,
+}
+
+impl PrioritizationFeeDetails {
+    pub fn new(fee_type: PrioritizationFeeType, max_compute_units: u64) -> Self {
+        let mut compute_ticks = max_compute_units
+            .saturating_div(COMPUTE_UNIT_TICK_SIZE)
+            .max(1);
+        match fee_type {
+            PrioritizationFeeType::Deprecated(fee) => Self {
+                fee,
+                priority: fee.saturating_div(compute_ticks),
+            },
+            PrioritizationFeeType::Rate(fee_rate) => {
+                if compute_ticks * COMPUTE_UNIT_TICK_SIZE < max_compute_units {
+                    compute_ticks = compute_ticks.saturating_add(1);
+                }
+
+                Self {
+                    fee: fee_rate.saturating_mul(compute_ticks),
+                    priority: fee_rate,
+                }
+            }
+        }
```

Comment:
This is the most important part of the patch. It converts from a fee into a fee rate and it converts from a fee rate into a prioritization fee. It also handles the case when a tx requests 0 compute units.